src/IO/ITKImageInputFileFormat.cxx

```diff
@@ -81,46 +109,124 @@ ITKImageInputFileFormat::can_read(const FileSignature& /*signature*/,
+template<typename STIRImageType>
+unique_ptr<STIRImageType>
+ITKImageInputFileFormat<STIRImageType>::
+read_from_file(const std::string& filename) const
 {
   return
-    unique_ptr<DiscretisedDensity<3,float> >
-    (read_file_itk< FinalImageType >(filename));
+    unique_ptr<STIRImageType>
+    (read_file_itk< STIRImageType >(filename));
+}
+
+// Actual conversion function
+STIRImageSingle*
+convert_ITK_to_STIR(const ITKImageSingle::Pointer itk_image_orig)
+{
+    ITKImageSingle::Pointer itk_image;
+    CartesianCoordinate3D<float> voxel_size;
+    BasicCoordinate<3,int> min_indices, max_indices;
+    CartesianCoordinate3D<float> origin;
+
+    // orientate the ITK image
+    orient_ITK_image<ITKImageSingle>(
+                itk_image, voxel_size, min_indices, max_indices, origin, itk_image_orig);
+
+    // create STIR image
+    VoxelsOnCartesianGrid<float>* image_ptr =
+        new VoxelsOnCartesianGrid<float>
+            (IndexRange<3>(min_indices, max_indices),
+            origin,
+            voxel_size);
+
+    // copy data
+    VoxelsOnCartesianGrid<float>::full_iterator stir_iter = image_ptr->begin_all();
+    typedef itk::ImageRegionConstIterator< ITKImageSingle > IteratorType;
+    IteratorType it (itk_image, itk_image->GetLargestPossibleRegion() );
+    for ( it.GoToBegin(); !it.IsAtEnd(); ++it, ++stir_iter  ) {
+        *stir_iter = static_cast<float>(it.Get());
+    }
+
+    return image_ptr;
 }
 
+// Actual conversion function
+STIRImageMulti*
+convert_ITK_to_STIR(const ITKImageMulti::Pointer itk_image_orig)
+{
+    ITKImageMulti::Pointer itk_image;
+    CartesianCoordinate3D<float> voxel_size;
+    BasicCoordinate<3,int> min_indices, max_indices;
+    CartesianCoordinate3D<float> origin;
+
+    // orientate the ITK image
+    orient_ITK_image<ITKImageMulti>(
+                itk_image, voxel_size, min_indices, max_indices, origin, itk_image_orig);
+
+    // create STIR image
+    STIRImageMulti* image_ptr =
+        new STIRImageMulti
+            (IndexRange<3>(min_indices, max_indices),
+             origin,
+             voxel_size);
+
+    // copy data
+    STIRImageMulti::full_iterator stir_iter = image_ptr->begin_all();
+    typedef itk::ImageRegionConstIterator< ITKImageMulti > IteratorType;
+    IteratorType it (itk_image, itk_image->GetLargestPossibleRegion() );
+    for ( it.GoToBegin(); !it.IsAtEnd(); ++it, ++stir_iter) {
+        itk::Point<double,3U> itk_coord;
+        itk_coord[0] = -it.Get()[0]; // Seems we need an extra flip in the x-axis. don't know why...
+        itk_coord[1] = -it.Get()[1]; // Seems we need an extra flip in the y-axis. don't know why...
+        itk_coord[2] =  it.Get()[2];
+        *stir_iter =
+            ITK_coordinates_to_STIR(
+                itk_coord,
+                voxel_size,
+                min_indices,
+                true);
+    }
+
+    return image_ptr;
+}
 
 //To read any file format via ITK
-template<typename TImageType> VoxelsOnCartesianGrid<float>*
+template<>
```

Comment:
I'd rather stick to a single templated function (would need 2 template args, the ITK and STIR image types). see comments for the multi-version below on how that could work.

Reply:
Maybe I've misunderstood, but I used specialisations since methods of reading `single` and `multi` were very different.

Also, if I templated this method (`read_file_itk`) wrt `ITKImageType`, then I would also have to put the ITK type into the template of `read_from_file`. But I can't do that, since `read_from_file` is in the registeries. Am I wrong?

Reply:
There is no technical problem to template `ITKImageInputFileFormat` in both STIR and ITK image type. The registry is STIR image type specific, but the derived classes can have any type whatsoever, e.g.
```c++
RegisterInputFileFormat<ITKImageInputFileFormat<
         VoxelsOnCartesianGrid<CartesianCoordinate3D<float> >, 
         somefunkyITKtype, someothertypes > >
```
question is if the instantiation are really different. I'm not sure. In any case `read_file_itk` should be a `static` function

Reply:
They are quite different, so I don't think they should be combined.

Ok for making the function prototype `static` (apparently not allowed for specialisations). Just for clarification, were you asking me to do this because you thought `read_file_itk` was a member function? (Because it's not.)

Reply:
`read_file_itk` made static means it's a function local to the .cxx (not exposed to the linker), which is very different from a static member function. Sadly, `static` is an overloaded C++ keyword.

Fine to leave the 2 instantiations for `read_file_itk`

Reply:
I've made them static